### ISSUE TITLE
uid/gid are not set to 0 when run as non-root user

### ISSUE THREAD

**Issue by ShadowNinja** [archlinux/mkinitcpio#57]
mkinitcpio includes a pipeline like the following to generate the initcpio:
`bsdtar --null -cnf - -T - | bsdtar --uid 0 --gid 0 --null -cf - --format=newc @-`

It attempts to force the uid/gid to 0, but it doesn't actually work.  If you run it as a non-root user and then run `bsdtar -tvf initramfs.img` you'll see that the uid/gid is set to your uid/gid and you won't be able to boot with the image.

It seems like the issue is that bsdtar preserves uid/gid from tar files that are provided as input (using `@file` syntax) and doesn't apply the `--uid` or `--gid` options to those files.

Moving the `--uid 0 --gid 0` options to the first `bsdtar` command instead of the second one fixes the issue.

(also the `--null` on the second bsdtar seems superfluous given that it's not taking a list of paths as input)

Demo:
```
$ mkdir test && cd test
$ touch -d '@0' test
$ ls | bsdtar -cnf - -T - | bsdtar --uid 0 --gid 0 -cf - --format=newc @- | bsdtar -tvf -
-rw-r--r--  0 1000   1000        0 Dec 31  1969 test
$ ls | bsdtar --uid 0 --gid 0 -cnf - -T - | bsdtar -cf - --format=newc @- | bsdtar -tvf -
-rw-r--r--  0 0      0           0 Dec 31  1969 test
```

**Comment by Foxboron** [comment]
The issue here is that it's there to ensure reproducible images. If that entails that image generation as your personal user is broken then that is a trade-off but should probably error early instead of producing invalid images.

With the proposed change, is the images still reproducible?

**Comment by ShadowNinja** [comment]
The issue right now is that the `--uid`/`--gid` options don't do anything.  Moving those options to the previous invocation of `bsdtar` would fix that.  I tested that this doesn't change the output with the following:

```bash
$ echo 'MODULES=()
BINARIES=()
FILES=()
HOOKS=(systemd autodetect modconf block sd-vconsole sd-encrypt lvm2 filesystems keyboard fsck)' > mkinitcpio.conf
$ sudo mkinitcpio -c mkinitcpio.conf -g root-before.img
$ mkinitcpio -c mkinitcpio.conf -g unpriv-before.img
$ # Now apply patch below to mkinitcpio
$ sudo mkinitcpio -c mkinitcpio.conf -g root-after.img
$ mkinitcpio -c mkinitcpio.conf -g unpriv-after.img
$ sha256sum *.img
$ # All checksums match except unpriv-before.img
```

```diff
@@ -224,8 +224,8 @@
     # If this pipeline changes, |pipeprogs| below needs to be updated as well.
     find . -mindepth 1 -printf '%P\0' |
             sort -z |
-            LANG=C bsdtar --null -cnf - -T - |
-            LANG=C bsdtar --uid 0 --gid 0 --null  -cf - --format=newc @- |
+            LANG=C bsdtar --uid 0 --gid 0 --null -cnf - -T - |
+            LANG=C bsdtar -cf - --format=newc @- |
             $compress "${COMPRESSION_OPTIONS[@]}" > "$out"
 
     pipestatus=("${PIPESTATUS[@]}")
```

I also removed the second `--null` since that also doesn't do anything since it's not taking in a list of files as input.

**Comment by grazzolini** [comment]
Fixed by #38
